edb/pgsql/metaschema.py

```diff
@@ -1836,6 +1839,115 @@ def __init__(self) -> None:
+    text = r'''
+        SELECT concat('\'',
+            replace(
+                replace(val, '\\', '\\\\'),
+                '\'', '\\\''),
+            '\'')
+    '''
+
+    def __init__(self) -> None:
+        super().__init__(
+            name=('edgedb', 'quote_literal'),
+            args=[('val', ('text',))],
+            returns=('str',),
+            volatility='immutable',
+            text=self.text)
+
+
+class ConfigObjectAsDDLFunction(dbops.Function):
+    """Describe single configuration object as DDL"""
+
+    def __init__(self, compiler, schema, name: str, plural: str) -> None:
+
+        cfg = schema.get('cfg::' + name)
+        items = []
+        for pn, p in cfg.get_pointers(schema).items(schema):
+            if pn in ('id', '__type__'):
+                continue
+            if p.get_annotation(schema, 'cfg::internal') == 'true':
+                continue
+            ptype = p.get_target(schema)
+            mult = p.get_cardinality(schema) is qltypes.SchemaCardinality.MANY
+            if isinstance(ptype, s_objtypes.ObjectType):
+                items.append(
+                    f"'  { qlquote.quote_ident(pn) } := (INSERT ' ++ "
+                    f" cfg::{ name }.{ qlquote.quote_ident(pn) }.__type__.name"
+                    f" ++ '),\\n'"
+                )
+            elif mult:
+                items.append(
+                    f"'  ' ++ cfg::_name_value({ ql(pn) },"
+                    f" array_agg(cfg::{ name }.{ qlquote.quote_ident(pn) }))"
+                    f" ++ ',\\n'")
+            else:
+                items.append(
+                    f"'  ' ++ cfg::_name_value({ ql(pn) },"
+                    f" cfg::{ name }.{ qlquote.quote_ident(pn) })"
+                    f" ++ ',\\n'")
+
+        script = (
+            f"SELECT array_join(array_agg(("
+            f"SELECT 'CONFIGURE SYSTEM INSERT {name} {{\\n'"
+            f" ++ {' ++ '.join(items)} ++ '}};\n'"
+            f")), '')"
+        )
+        _, text = edbbootstrap.compile_bootstrap_script(
+            compiler, schema, script,
+            output_format=edbcompiler.IoFormat.BINARY)
+        super().__init__(
+            name=('edgedb', '_config_insert_all_'+plural),
+            args=[],
+            returns=('text'),
+            # Stable because it's raising exceptions.
+            volatility='stable',
+            text=text)
+
+
+class DescribeSystemConfigAsDDLFunction(dbops.Function):
+    """Describe server configuration as DDL"""
+
+    def __init__(self, compiler, schema) -> None:
+
+        cfg = schema.get('cfg::Config')
+        items = []
+        for pn, p in cfg.get_pointers(schema).items(schema):
+            if pn in ('id', '__type__'):
+                continue
+            ptype = p.get_target(schema)
+            mult = p.get_cardinality(schema) is qltypes.SchemaCardinality.MANY
+            if isinstance(ptype, s_objtypes.ObjectType):
+                if pn not in {'sessobj', 'sysobj'}:
```

Comment:
These are test config objects that only get populated in dev mode.  There's no need to filter them out in general.

Reply:
I don't think we should really care about devmode vs non-devmode scenario.  The worst case that would happen is that you'll have a couple of extra definitions in `cfg`, no big deal.  On the other hand, we will certainly be adding testing stuff to `cfg`, and it's easy to overlook this filter here, making it incomplete and moot.

Reply:
I still don't understand what to do:
1. These are objects and currently each object have a special treatment. I do know that to do with `auth` and `ports` but don't how to serialize `sessobj` and `sysobj`. Also If I understand correctly `sessobj` has something to the current session, so does not need to be in dump regardless of the dev/non-dev mode.
2. Worst case is dump that has to be edited manually, not "extra records that are silently ignored"

Reply:
OK, let's keep the check as is to progress.  Ideally, we'd need to generalize the handling of nested config objects to avoid having to add code here when we add stuff to `cfg::` (i.e not hardcode `Port` and `Auth`, but iterate over `cfg::Config` links instead).  You actually don't need to declare subclasses `dbops.Function` to declare functions, you can construct a `dbops.Function` directly in an introspection loop in `generate_more_support_functions`.